### PR TITLE
Cancel statement on completion of `execute-reducible-query`

### LINKED ISSUE #39018: Metabase renders the result but the query does not finish in the DB
### Describe the bug

Queries in the DB seem to continue in the background, while the query has already been finished for Metabase

### To Reproduce

1) add the Redshift dev DW to a Metabase (sample DB)
2) do the following query:
```
SELECT
  "public"."random_data"."text_column1" AS "text_column1",
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0 AS "numeric_column1",
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0 AS "numeric_column4",
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0 AS "numeric_column5",
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0 AS "numeric_column8",
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0 AS "numeric_column9",
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0 AS "numeric_column10",
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0 AS "numeric_column13",
  COUNT(*) AS "count"
FROM
  "public"."random_data"
GROUP BY
  "public"."random_data"."text_column1",
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0,
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0,
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0,
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0
ORDER BY
  "public"."random_data"."text_column1" ASC,
  FLOOR(("public"."random_data"."numeric_column1" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column4" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column5" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column8" / 125.0)) * 125.0 ASC,
  FLOOR(("public"."random_data"."numeric_column9" / 125.0)) * 125.0 ASC,
  FLOOR(
    ("public"."random_data"."numeric_column10" / 125.0)
  ) * 125.0 ASC,
  FLOOR(
    ("public"."random_data"."numeric_column13" / 125.0)
  ) * 125.0 ASC
```
3) wait till it finishes and then...
4) go to Redshift and check the status of `select * from STV_INFLIGHT`, you'll see that the query is still running while in Metabase it has finished

### Expected behavior

The connection and query should finish if Metabase finishes

### Logs

NA

### Information about your Metabase installation

```JSON
v48.6
```

### Severity

P1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lbrdnk**
Fixes #39018, #41448

This PR adds call to `.cancel` after `respond` completion in `metabase.driver.sql-jdbc.execute/execute-reducible-query`. 

That ensures statements are not in progress prior to `.close` calls of `with-open` of that function. In case statement was still in progress, the close call would block.

That situation could arise if `limit` (as per `metabase.query-processor.middleware.limit/limit-xform` middleware) takes the necessary results subset before statement has completed.

---

I do not add new tests for the change, as I believe it is overkill to check that query was actually canceled per every driver. That should be guaranteed by java driver implementation. Exception is handled for the case, where the method is not implemented by a java driver.

### CODE REVIEW

File: src/metabase/driver/sql_jdbc/execute.clj [hunk 1]
```diff
@@ -710,7 +710,18 @@
                                                     e))))]
         (let [rsmeta           (.getMetaData rs)
               results-metadata {:cols (column-metadata driver rsmeta)}]
-          (respond results-metadata (reducible-rows driver rs rsmeta qp.pipeline/*canceled-chan*))))))))
+          (try (respond results-metadata (reducible-rows driver rs rsmeta qp.pipeline/*canceled-chan*))
+               ;; Following cancels the statment on the dbms side.
+               ;; It avoids blocking `.close` call, in case we reduced the results subset eg. by means of
```
Comment (metamben) [inline src/metabase/driver/sql_jdbc/execute.clj:715]: I think we should mention that this a problem specific to redshift reference the bug ticker.

File: src/metabase/driver/sql_jdbc/execute.clj [hunk 2]
```diff
@@ -710,7 +710,18 @@
...
+               ;; [[metabase.query-processor.middleware.limit/limit-xform]] middleware, while statment is still
+               ;; in progress.
+               ;; It also handles situation where query is canceled through [[qp.pipeline/*canceled-chan*]].
+               (finally (try (.cancel stmt)
+                             (catch java.sql.SQLException _
+                               (log/warn "Statement cancelation failed."))
+                             (catch java.sql.SQLFeatureNotSupportedException _
+                               (log/warnf "Statemet's `.cancel` method not supported by `%s` driver."
```
Comment (metamben) [inline src/metabase/driver/sql_jdbc/execute.clj:723]:
```suggestion
                               (log/warnf "Statemet's `.cancel` method not supported by the `%s` driver."
```

File: src/metabase/driver/sql_jdbc/execute.clj [hunk 3]
```diff
@@ -710,7 +710,18 @@
...
+                                          (name driver))))))))))))
```
Comment (metamben) [inline src/metabase/driver/sql_jdbc/execute.clj:724]: I suggest catching any other exceptions too.

**Review by metamben — APPROVED** [review]
LGTM

**Comment by lbrdnk** [comment]
@metabase-bot backport release-x.48.x

**Pull request merged**